---
layout: welcome
title: Publications
buttons:
  print: false
  pdf: /assets/Publication_List_Kieran_Didi.pdf
---
<!-- PDF version available [here]({{ site.baseurl }}/assets/Publication_List_Kieran_Didi.pdf){:.no-push-state}. -->
## Publications 

### Pre-Prints
 - La-Proteina: Atomistic Protein Generation via Partially Latent Flow Matching
  Tomas Geffner\*, **Kieran Didi**\*, Danny Reidenbach, Zhonglin Cao, Zuobai Zhang, Christian Dallago, Emine Kucukbenli, Arash Vahdat, Karsten Kreis, **arxiv**, 07/2025
  <br>
  <div class="publication-buttons">
    <a href="https://arxiv.org/abs/2507.09466" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
    <a href="https://github.com/NVIDIA-Digital-Bio/la-proteina" class="pub-button code" target="_blank" rel="noopener">ğŸ’» Code</a>
    <a href="https://research.nvidia.com/labs/genair/la-proteina/" class="pub-button webpage" target="_blank" rel="noopener">ğŸŒ Webpage</a>
  </div>

- Highly efficient protein structure prediction on NVIDIA RTX Blackwell and Grace-Hopper
  **Kieran Didi\***, Prashant Sohani\*, Fabian Berressem, Alexander Nesterovskiy, Boris Fomitchev, Robert Ohannessian, Mohamed Elbalkini, Jonathan Cogan, Anthony Costa, Arash Vahdat, Felix Kallenborn, Bertil Schmidt, Milot Mirdita, Martin Steinegger, Christian Dallago, Alejandro Chacon, **Technical Report**, 09/2025
  <br>
  <div class="publication-buttons">
    <a href="https://research.nvidia.com/labs/dbr/assets/data/manuscripts/fast_folding.pdf" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
    <a href="https://developer.nvidia.com/blog/accelerate-protein-structure-inference-over-100x-with-nvidia-rtx-pro-6000-blackwell-server-edition/" class="pub-button blog" target="_blank" rel="noopener">ğŸ’¬ Blog</a>

- De novo Design of All-atom Biomolecular Interactions with RFdiffusion3
  Jasper Kenneth Veje Butcher, Rohith Krishna, Raktim Mitra, Rafael Isaac Brent, Yanjing Li, Nathaniel Corley, Paul Kim, Jonathan Funk, Simon Valentin Mathis, Saman Salike, Aiko Muraishi, Helen Eisenach, Tuscan Rock Thompson, Jie Chen, Yuliya Politanska, Enisha Sehgal, Brian Coventry, Odin Zhang, Bo Qiang, **Kieran Didi**, Maxwell Kazman, Frank DiMaio, David Baker, **biorxiv**, 09/2025
  <br>
  <div class="publication-buttons">
  <a href="https://www.biorxiv.org/content/10.1101/2025.09.18.676967v1" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
  </div>

- Accelerating Biomolecular Modeling with AtomWorks and RF3
  Nathaniel Corley\*, Simon Mathis\*, Rohith Krishna\*, Magnus S Bauer, Tuscan R Thompson, Woody Ahern, Maxwell W Kazman, Rafael I Brent, **Kieran Didi**, Andrew Kubaney, Lilian McHugh, Arnav Nagle, Andrew Favor, Meghana Kshirsagar, Pascal Sturmfels, Yanjing Li, Jasper Butcher, Bo Qiang, Lars L Schaaf, Raktim Mitra, Katelyn Campbell, Odin Zhang, Roni Weissman, Ian R Humphreys, Qian Cong, Jonathan Funk, Shreyash Sonthalia, Pietro LiÃ², David Baker, Frank DiMaio, **biorxiv**, 08/2025
  <br>
  <div class="publication-buttons">
    <a href="https://www.biorxiv.org/content/10.1101/2025.08.14.670328v2.abstract" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
    <a href="https://github.com/RosettaCommons/atomworks/tree/production" class="pub-button code" target="_blank" rel="noopener">ğŸ’» Code</a>

- MotifBench: A standardized protein design benchmark for motif-scaffolding problems
  Zhuoqi Zheng, Bo Zhang, **Kieran Didi**, Kevin K. Yang, Jason Yim, Joseph L. Watson, Hai-Feng Chen, Brian L. Trippe, **arxiv**, 02/2025
  <br>
  <div class="publication-buttons">
    <a href="https://arxiv.org/abs/2502.12479" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
    <a href="https://github.com/blt2114/MotifBench" class="pub-button code" target="_blank" rel="noopener">ğŸ’» Code</a>
  </div>

- BioNeMo Framework: a modular, high-performance library for AI model development in drug discovery
  Peter St. John et al, **arxiv**, 11/2024
  <br>
  <div class="publication-buttons">
    <a href="https://arxiv.org/abs/2411.10548" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
    <a href="https://github.com/BioNeMo/BioNeMo" class="pub-button code" target="_blank" rel="noopener">ğŸ’» Code</a>
    <a href="https://www.nvidia.com/en-us/clara/biopharma/" class="pub-button webpage" target="_blank" rel="noopener">ğŸŒ Webpage</a>
    <a href="https://nvidianews.nvidia.com/news/nvidia-opens-bionemo-to-scale-digital-biology-for-global-biopharma-and-scientific-industry" class="pub-button blog" target="_blank" rel="noopener">ğŸ’¬ Blog</a>
  </div>

- BioModelsML: Building a FAIR and reproducible collection of machine learning models in life sciences and medicine for easy reuse
  Divyang Deep Tiwari, Nils Hoffmann, **Kieran Didi**, Sumukh Deshpande, Sucheta Ghosh, Tung V. N. Nguyen, Karthik Raman, Henning Hermjakob, Rahuman S Malik Sheriff, **biorxviv**, 05/2023
  <br>
  <div class="publication-buttons">
    <a href="https://www.biorxiv.org/content/10.1101/2023.05.22.540599v1" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
    <a href="https://www.ebi.ac.uk/biomodels/" class="pub-button webpage" target="_blank" rel="noopener">ğŸŒ Webpage</a>
  </div>


### Science Journals
- GPU-accelerated homology search with MMseqs2
  Felix Kallenborn\*, Alejandro Chacon\*, Christian Hundt, Hassan Sirelkhatim, **Kieran Didi**, Sooyoung Cha, Christian Dallago, Milot Mirdita, Bertil Schmidt, Martin Steinegger, **Nature Methods**, 09/2025
  <br>
  <div class="publication-buttons">
    <a href="https://www.nature.com/articles/s41592-025-02819-8" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
    <a href="https://github.com/soedinglab/MMseqs2" class="pub-button code" target="_blank" rel="noopener">ğŸ’» Code</a>
    <a href="https://www.biorxiv.org/content/10.1101/2024.11.13.623350v4.abstract" class="pub-button webpage" target="_blank" rel="noopener">ğŸŒ Webpage</a>
    <a href="https://developer.nvidia.com/blog/boost-alphafold2-protein-structure-prediction-with-gpu-accelerated-mmseqs2/" class="pub-button blog" target="_blank" rel="noopener">ğŸ’¬ Blog</a>
    <a href="https://www.biorxiv.org/content/10.1101/2024.11.13.623350v4.abstract" class="pub-button poster" target="_blank" rel="noopener">Poster</a>
    <a href="https://www.biorxiv.org/content/10.1101/2024.11.13.623350v4.abstract" class="pub-button slides" target="_blank" rel="noopener">ğŸ“„ Slides</a>
    <a href="https://www.biorxiv.org/content/10.1101/2024.11.13.623350v4.abstract" class="pub-button video" target="_blank" rel="noopener">ğŸ¥ Video</a>
  </div>
  
- Structure-based Drug Design with Equivariant Diffusion Models
  Arne Schneuing\*, Yuanqi Du\*, Charles Harris\*, **Kieran Didi**, Arian Jamasb, Ilia Igashov, weitao Du, Tom Blundell, Pietro LiÃ², Carla Gomes, Max Welling, Michael Bronstein, Bruno Correia, **Nature Computational Science**, 12/2024
  <br>
  <div class="publication-buttons">
    <a href="https://www.nature.com/articles/s43588-024-00737-x" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
    <a href="https://github.com/arneschneuing/DiffSBDD" class="pub-button code" target="_blank" rel="noopener">ğŸ’» Code</a>
  </div>

- MISATO: machine learning dataset of proteinâ€“ligand complexes for structure-based drug discovery
  Till Siebenmorgen\*, Filipe Menezes\*, Sabrina Benassou, Erinc Merdivan, **Kieran Didi**, AndrÃ© Santos Dias MourÃ£o, RadosÅ‚aw Kitel, Pietro LiÃ², Stefan Kesselheim, Marie Piraud, Fabian J. Theis, Michael Sattler & Grzegorz M. Popowicz, **Nature Computational Science**, 05/2024
  <br>
  <div class="publication-buttons">
    <a href="https://www.nature.com/articles/s43588-024-00627-2" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
    <a href="https://github.com/t7morgen/misato-dataset" class="pub-button code" target="_blank" rel="noopener">ğŸ’» Code</a>
    <a href="https://www.nature.com/articles/s43588-024-00631-6" class="pub-button press" target="_blank" rel="noopener">Press</a>
  </div>

- Synsor: a tool for alignment-free detection of engineered DNA sequences
  Aidan P Tay, **Kieran Didi**, Anuradha Wickramarachchi, Denis Bauer, Laurence Wilson and Maciej Maselko, **Frontiers in Bioengineering and Biotechnology**, 04/2024
  <br>
  <div class="publication-buttons">
    <a href="https://www.frontiersin.org/journals/bioengineering-and-biotechnology/articles/10.3389/fbioe.2024.1375626/full" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
    <a href="https://github.com/aidantay/Synsor" class="pub-button code" target="_blank" rel="noopener">ğŸ’» Code</a>
  </div>

- AbNatiV: VQ-VAE-based assessment of antibody and nanobody nativeness for hit selection, humanisation, and engineering
  Aubin Ramon, Montader Ali, Misha Atkinson, Alessio Saturnino, **Kieran Didi**, Cristina Visentin, Stefano Ricagno, Xing Xu, Matthew Greenig, Pietro Sormanni, **Nature Machine Intelligence**, 04/2023
  <br>
  <div class="publication-buttons">
    <a href="https://www.nature.com/articles/s42256-023-00778-3" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
    <a href="https://gitlab.developers.cam.ac.uk/ch/sormanni/abnativ" class="pub-button code" target="_blank" rel="noopener">ğŸ’» Code</a>
  </div>

- Biomolecular condensate phase diagrams with a combinatorial microdroplet platform
  William E. Arter\*, Runzhang Qi\*, Nadia A. Erkamp\*, Georg Krainer\*, **Kieran Didi**, Timothy J. Welsh, Julia Acker, Jonathan Nixon-Abell, Seema Qamar, Jordina GuillÃ©n-Boixet, Titus M. Franzmann, David Kuster, Anthony A. Hyman, Alexander Borodavka, Peter St George-Hyslop, Simon Alberti & Tuomas P. J. Knowles, **Nature Communications**, 12/2022
  <br>
  <div class="publication-buttons">
    <a href="https://www.nature.com/articles/s41467-022-35265-7" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
  </div>


### ML Conferences (Main Track)
- [Compositional Flows for 3D Molecule and Synthesis Pathway Co-design](https://arxiv.org/abs/2504.08051), Tony Shen\*, Seonghwan Seo\*, Ross Irwin, **Kieran Didi**, Simon Olsson, Woo Youn Kim, Martin Ester, **ICML 2025**
  <br>
  <div class="publication-buttons">
    <a href="https://arxiv.org/abs/2504.08051" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
  </div>

- [Proteina: Scaling Flow-based Protein Structure Generative Models](https://arxiv.org/abs/2503.00710), Tomas Geffner\*, **Kieran Didi**\*, Zuobai Zhang\*, Danny Reidenbach, Zhonglin Cao, Jason Yim, Mario Geiger, Christian Dallago, Emine Kucukbenli, Arash Vahdat, Karsten Kreis, **ICLR 2025 (oral presentation)**
  <br>
  <div class="publication-buttons">
    <a href="https://arxiv.org/abs/2503.00710" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
    <a href="https://github.com/NVIDIA-Digital-Bio/proteina/" class="pub-button code" target="_blank" rel="noopener">ğŸ’» Code</a>
    <a href="https://research.nvidia.com/labs/genair/proteina/" class="pub-button webpage" target="_blank" rel="noopener">ğŸŒ Webpage</a>
    <a href="https://www.youtube.com/watch?v=Y2dRj9_ZEHw&pp=ygUWcHJvdGVpbmEga2Fyc3RlbiBrcmVpcw%3D%3D" class="pub-button video" target="_blank" rel="noopener">ğŸ¥ Video</a>

  </div>

- [DEFT: Efficient Finetuning of Conditional Diffusion Models by Learning the Generalised h-transform](https://arxiv.org/abs/2406.01781), Alexander Denker\*, Francisco Vargas\*, Shreyas Padhy\*, **Kieran Didi**\*, Simon Mathis\*, Vincent Dutordoir, Riccardo Barbano, Emile Mathieu, Urszula Julia Komorowska, Pietro Lio, **NeurIPS 2024**
  <br>
  <div class="publication-buttons">
    <a href="https://arxiv.org/abs/2406.01781" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
    <a href="https://www.youtube.com/watch?v=GrtWwD8pPWY" class="pub-button video" target="_blank" rel="noopener">ğŸ¥ Video</a>

  </div>

- [Dynamics-Informed Protein Design with Structure Conditioning](https://openreview.net/forum?id=jZPqf2G9Sw), Urszula Julia Komorowska\*, Simon V Mathis\*, **Kieran Didi**, Francisco Vargas, Pietro Lio, Mateja Jamnik , **ICLR 2024**
  <br>
  <div class="publication-buttons">
    <a href="https://openreview.net/forum?id=jZPqf2G9Sw" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
  </div>

- [Evaluating Representation Learning on the Protein Structure Universe](https://arxiv.org/abs/2406.13864), Arian Rokkum Jamasb\*, Alex Morehead\*, Zuobai Zhang\*, Chaitanya Joshi\*, **Kieran Didi**, Simon Mathis, Charles Harris, Jian Tang, Jianlin Cheng, Pietro LiÃ², Tom Blundell, **ICLR 2024**
  <br>
  <div class="publication-buttons">
    <a href="https://arxiv.org/abs/2406.13864" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
    <a href="https://github.com/a-r-j/ProteinWorkshop" class="pub-button code" target="_blank" rel="noopener">ğŸ’» Code</a>
  </div>


### ML Conferences (Workshop Track)
- [RNA-FrameFlow: Flow Matching for de novo 3D RNA Backbone Design](https://arxiv.org/abs/2406.13839), Rishabh Anand\*, Chaitanya K. Joshi\*, Alex Morehead, Arian R. Jamasb, Charles Harris, Simon V. Mathis, **Kieran Didi**, Bryan Hooi, Pietro LiÃ², **ICML SPIGM 2024 (Oral Presentation)**, 07/2024
  <br>
  <div class="publication-buttons">
    <a href="https://arxiv.org/abs/2406.13839" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
    <a href="https://github.com/rish-16/rna-backbone-design" class="pub-button code" target="_blank" rel="noopener">ğŸ’» Code</a>
  </div>

- [A framework for conditional diffusion modelling with applications in motif scaffolding for protein design](https://arxiv.org/abs/2312.09236), **Kieran Didi***, Francisco Varga*, Simon Mathis*, Vincent Dutordoir*, Emile Mathieu, Urszula Julia Komorowska, Pietro Lio, **NeurIPS AI4D3 2023 (Oral Presentation)**, 11/2023
  <br>
  <div class="publication-buttons">
    <a href="https://arxiv.org/abs/2312.09236" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
  </div>

- [Benchmarking Generated Poses: How Rational is Structure-based Drug Design with Generative Models?](https://arxiv.org/abs/2308.07413), Charles Harris, **Kieran Didi**, Arian Jamasb, Chaitanya Joshi, Simon Mathis, Pietro Lio, Tom Blundell, **NeurIPS MLSB 2023 (Oral Presentation)**, 11/2023
  <br>
  <div class="publication-buttons">
    <a href="https://arxiv.org/abs/2308.07413" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
    <a href="https://github.com/cch1999/posecheck" class="pub-button code" target="_blank" rel="noopener">ğŸ’» Code</a>
  </div>

- Modelling biology in novel ways - an AI-first course in Structural Bioinformatics, **Kieran Didi**, Charles Harris, Pietro LiÃ², Rainer Beck, **NeurIPS AI4Science 2023**, 10/2023 
  <br>
  <div class="publication-buttons">
    <a href="#" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
  </div>

- [Flexible Small-Molecule Design and Optimization with Equivariant Diffusion Models](https://drive.google.com/file/d/11kSYs6WYAg2_D0HtF8NxG7e6dUGQcVaL/view), Charles Harris, **Kieran Didi**, Arne Schneuing, Yuanqi Du, Arian Jamasb, Michael Bronstein, Bruno Correia, Pietro LiÃ², Tom Blundell, **ICLR MLDD 2023**, 03/2023
  <br>
  <div class="publication-buttons">
    <a href="https://drive.google.com/file/d/11kSYs6WYAg2_D0HtF8NxG7e6dUGQcVaL/view" class="pub-button paper" target="_blank" rel="noopener">ğŸ“„ Paper</a>
  </div>

### Book Chapters

- *Generative Models in Bioinformatics*, Mitchell J O'Brien, Letitia MF Sng, Priya Ramarao-Milne, **Kieran Didi**, Denis C Bauer, **Encyclopedia of Bioinformatics and Computational Biology**, 01/2023
  <br>
  <div class="publication-buttons">
  </div>

## Conference Presentations
- [Why equivariance is (not) so popular in Life and Material Science](https://ml4lms.bio/blog/1250) (invited talk), ICML ML4LMS 2024, 07/2024
- [Conditioning generative models for proteins, images and more](https://www.embl.org/about/info/course-and-conference-office/events/ees24-01/) (oral presentation), AI and Biology Conference, EMBL Heidelberg (Germany), 03/2024
- [Where do we stand in protein and ligand design?](https://www.embl.org/about/info/course-and-conference-office/events/csb23-01/) (oral presentation), Computational Structural Biology Conference, EMBL Heidelberg (Germany), 12/2023
- [Conditional Diffusion Models for Protein Design via Doob's h-transform](https://www.moml.mit.edu/) (poster), MoML Conference, MIT (US), 11/2023
- [Reality Check - Evaluating Generative Models for Protein and Ligand Design](https://www.broadinstitute.org/machine-learning-drug-discovery-symposium/machine-learning-drug-discovery-symposium) (poster), Machine Learning in Drug Discovery Symposium, Broad Institute (US), 10/2023
- [Accessible AMR-detection tools for clinicians via cloud-based bioinformatics](https://pheedloop.com/ABACBS2022/site/CI) (oral presentation), Clinical Informatics Symposium, Melbourne (Australia), 12/2022
- [sINSIDER: a cloud-native modular platform for kmer-based genomic analysis](https://www.abacbs.org/conference2022) (oral presentation), ABACBS Conference, Melbourne (Australia), 11/2022
- [Moving bioinformatics to the cloud](https://escamps.org/) (keynote presentation), eSCAMPS Symposium, Cambridge (UK), 09/2022
- [Detecting DNA integration via HPC bioinformatics pipeline](https://na.eventscloud.com/website/36005/home/) (poster),R&I Conference, Sydney (Australia), 08/2022

## Reviewing

ICML CompBio 2023, NeurIPS AI4Science 2023, NeurIPS GenBio 2023, MoML@MIT 2023, ICLR GEM 2024, NeurIPS 2025

## Teaching
- [Deep Learning in Structure-Based Drug Design](https://www.sib.swiss/training/course/20240609_CADD), tutorial, Spring School Structure-based Computer-aided Drug Design, Swiss Institute of Bioinformatics (SIB), 06/2024
- [SDEs and Diffusion Processes](https://sde-course.netlify.app/), lecture, Cambridge University (MPhil Advanced Computer Science), 01/2024
- [Structural Bioinformatics](https://structural-bioinformatics.netlify.app/), lecture, Heidelberg University (BSc Biochemistry), 10/2023
- [Graph Modelling with Applications to Structural Biology](https://www.vanderschaar-lab.com/ccaim-ai-and-machine-learning-summer-school/), presentation, CCAIM AI and Machine Learning Summer School, 09/2023
- [Advanced Python for Scientists](https://github.com/kierandidi/advanced_python_for_scientists), lecture, Heidelberg University, 06/2023
- [Algorithms II](https://www.cl.cam.ac.uk/teaching/2223/Algorithm2/), supervising course on data structures and graph algorithms, Cambridge University, 03-05/2023
- [Algorithms I](https://www.cl.cam.ac.uk/teaching/2223/Algorithm1/), Cambridge University, 01-03/2023
- *MLOps with PyTorch Lightning and W&B*, workshop, ML Forum CSIRO Australia, 01/2023
- [Python for Scientists](https://github.com/kierandidi/python_for_scientists), lecture, Heidelberg University, 11/2022
- [Python Best Practices](https://ssciwr.github.io/Python-best-practices-course/), workshop (co-host), Scientific Software Center Heidelberg, 11/2022
- [Deep Learning with PyTorch](https://ssciwr.github.io/lunch-time-python/#:~:text=Lunch%20Time%20Python%20aims%20at,will%20be%20made%20available%20afterwards.), Lunch Time Python workshop, Scientific Software Center Heidelberg, 10/2022
- [Python for Biochemists](https://github.com/kierandidi/Python_for_Biochemists), lecture, Heidelberg University, 03/2022

